Fix fork tests

protolambda opened the pull request:
**Description**

This fixes our e2e action tests to properly configure up to fork N, and refactors the user txs test to run through all hardfork scenarios instead of just regolith and ecotone.

With this, I caught an issue in the ecotone receipts RPC: https://github.com/ethereum-optimism/op-geth/pull/219

protolambda committed:
op-e2e: fix fork tests

go.mod: update to include ecotone l1 fee scalar receipts nil check

review fixes

ajsutton commented on go.mod:
```diff
@@ -219,7 +219,7 @@ require (
 )
 
-replace github.com/ethereum/go-ethereum v1.13.5 => github.com/ethereum-optimism/op-geth v1.101305.2-rc.2
+replace github.com/ethereum/go-ethereum v1.13.5 => github.com/ethereum-optimism/op-geth v1.101305.2-rc.2.0.20240116235610-75ee01e84295
```
We'll need to point this to the commit that winds up on the optimism branch (hopefully it's a fast forward merge and so is the same but will verify once it has).

protolambda replied:
updated

ajsutton commented on op-e2e/actions/user_test.go:
```diff
@@ -30,42 +51,75 @@ type hardforkScheduledTest struct {
+			t.Run("at_genesis", func(t *testing.T) {
+				tc := hardforkScheduledTest{}
+				for _, f := range forks[:i+1] { // activate, all up to and incl this fork, at genesis
+					*tc.fork(f) = new(hexutil.Uint64)
```
That hurts my head a bit. Maybe it's worth `fork` accepting the block the fork should activate at and actually setting the variable instead of just returning a reference to it?

or possibly just have a `scheduleFork` method that does that if it's useful to be able to get a reference to the fork variable.

protolambda replied:
changed it to use getters/setters

ajsutton reviewed (commented):
Looks good once we verify the op-geth hash.

ajsutton reviewed (approved):
I lost the gamble of which merge option would work best on op-geth side so would be ideal if we can update the op-geth version to `d5f142e54a0a943e9f28530f130d06128cfa2eb4` which is the latest on optimism. The referenced commit is now on the optimism branch now though so I'm happy to approve.

The lint error will need to be fixed up as well but that's just a minor unused field in the tests.

protolambda commented:
Implemented review suggestions, removed an incorrect require-is-ecotone that I added while debugging, and updated op-geth go.mod reference. Should be ready for merging now, will queue it up

protolambda commented on op-e2e/actions/user_test.go:
```diff
@@ -14,14 +15,42 @@ import (
+	return (*uint64)(*tc.fork(fork))
+}
+
+func (tc *hardforkScheduledTest) fork(fork string) **hexutil.Uint64 {
+	switch fork {
+	case "fjord":
+		return &tc.fjordTime
+	case "ecotone":
+		return &tc.ecotoneTime
+	case "delta":
+		return &tc.deltaTime
+	case "canyon":
+		return &tc.canyonTime
+	case "regolith":
+		return &tc.regolithTime
+	default:
+		panic(fmt.Errorf("unrecognized fork: %s", fork))
+	}
+}
```
I've hidden it with getters/setters, so it's more friendly to use than the double-pointer.

protolambda committed:
op-e2e: fix lint

trianglesphere commented:
I really like this change. op-e2e is still failing in TestFees b/c of regolith (I think)

protolambda committed:
op-e2e: start making L1-fees test a bit more compatible, but GPO contract still has shortcomings

protolambda committed:
Merge branch 'develop' into fix-fork-tests

protolambda commented:
Turns out the old TestFees op-e2e test only covers pre-regolith, and the GPO contract doesn't return post-regolith L1 fee data. The latest version does support Ecotone fees, but only after `setEcotone()`, which we don't prepare as part of Ecotone genesis yet (added a reference to this fees test to the relevant issue). So keeping the test pre-regolith-only for now, but added skipped cases for the post-regolith and ecotone cases.

protolambda commented on op-e2e/system_test.go:
```diff
@@ -1083,13 +1114,7 @@ func TestFees(t *testing.T) {
-		},
-		BedrockBlock: big.NewInt(0),
-	}
+	config := sys.L2Genesis().Config
```
it's not dead code
